src/widgets/survey/model/hooks/useAutoCompletion.ts

```diff
@@ -0,0 +1,205 @@
+import { useCallback, useRef } from 'react';
+
+import { useQueryClient } from '@tanstack/react-query';
+
+import {
+  CompleteEntityIntoUploadToQueue,
+  EntityPath,
+  EntityPathParams,
+  ProcessAutocompletion,
+  StoreProgress,
+} from '@app/abstract/lib';
+import {
+  AnswersQueueService,
+  QueueProcessingService,
+} from '@app/entities/activity';
+import { AppletModel } from '@app/entities/applet';
+import { LogTrigger } from '@app/shared/api';
+import {
+  Emitter,
+  Logger,
+  Mutex,
+  useAppDispatch,
+  useAppSelector,
+} from '@app/shared/lib';
+
+import {
+  CollectCompletionsService,
+  CollectCompletionOutput,
+  ConstructCompletionsService,
+} from '../services';
+
+type Result = {
+  process: ProcessAutocompletion;
+  completeEntityIntoUploadToQueue: CompleteEntityIntoUploadToQueue;
+  hasItemsInQueue: boolean;
+};
+
+export const AutoCompletionMutex = Mutex();
+
+const useAutoCompletion = (): Result => {
+  const mutex = useRef(AutoCompletionMutex).current;
+
+  const notCompletedEntities = useAppSelector(
+    AppletModel.selectors.selectNotCompletedEntities,
+  );
+
+  const storeProgress: StoreProgress = useAppSelector(
+    AppletModel.selectors.selectInProgressApplets,
+  );
+
+  const dispatch = useAppDispatch();
+
+  const queryClient = useQueryClient();
+
+  const hasItemsInQueue = (): boolean => {
+    return AnswersQueueService.getLength() > 0;
+  };
+
+  const createConstructService = useCallback(() => {
+    return new ConstructCompletionsService(
+      null,
+      queryClient,
+      storeProgress,
+      QueueProcessingService,
+      dispatch,
+    );
+  }, [dispatch, queryClient, storeProgress]);
+
+  const constructInternal = async (
+    collectOutput: CollectCompletionOutput,
+    constructService: ConstructCompletionsService,
+  ) => {
+    try {
+      if (collectOutput.completionType === 'intermediate') {
+        await constructService.constructForIntermediate({
+          ...collectOutput,
+          flowId: collectOutput.flowId!,
+        });
+      }
+
+      if (collectOutput.completionType === 'finish') {
+        await constructService.constructForFinish(collectOutput);
```

Comment:
Since `completionType` is a field of `collectOutput` and `collectOutput` is passed down to `constructService`, can `constructService` make the decision of what constructor (constructForIntermediate, constructForFinish) function to call internally?